keras_cv/datasets/pascal_voc/segmentation.py

```diff
@@ -0,0 +1,320 @@
+):
+    """Convert the original segmentation PNG file to proper encoding for tf.io API.
+
+    The original PNG was in a 2D (without channel dimension), which will get a wrong value
+    when directly read by tf.io API. In this function, the original image will be first
+    read with PIL.image API, expand the last dimention and save back by TF API. This will
+    ensure the converted image can be properly consumed by TF API.
+
+    If a file with `xxx_converted.png` already exists, and `use_cache` is True, the
+    conversion will be skipped.
+
+    Args:
+        segmentation_file_path: string, the original PNG file path from the dataset.
+        override: boolean, whether to use the existing cached file result.
+    Returns:
+        the file path for converted PNG file, it will be suffixed with '_converted'.
+    """
+    dir_and_file_name, _ = os.path.splitext(segmentation_file_path)
+    updated_file_path = dir_and_file_name + "_updated.png"
+    if os.path.exists(updated_file_path) and use_cache:
+        return updated_file_path
+
+    with tf.io.gfile.GFile(segmentation_file_path, "rb") as f:
+        mask = Image.open(f)
+
+    original_mask = np.expand_dims(np.asarray(mask), axis=-1)
+    # # Write the mask to PNG via the TF io API, so that it can be read properly.
+    encoded_png = tf.io.encode_png(original_mask)
+    tf.io.write_file(updated_file_path, encoded_png)
+
+
+def parse_single_image(image_file_path):
+    data_dir, image_file_name = os.path.split(image_file_path)
+    data_dir = os.path.normpath(os.path.join(data_dir, os.path.pardir))
+    image_id, _ = os.path.splitext(image_file_name)
+    class_segmentation_file_path = os.path.join(
+        data_dir, "SegmentationClass", image_id + ".png"
+    )
+    class_segmentation_file_path = _convert_pascal_voc_segmentation_mask_encoding(
+        class_segmentation_file_path
+    )
+    object_segmentation_file_path = os.path.join(
+        data_dir, "SegmentationObject", image_id + ".png"
+    )
+    object_segmentation_file_path = _convert_pascal_voc_segmentation_mask_encoding(
+        object_segmentation_file_path
+    )
+    annotation_file_path = os.path.join(data_dir, "Annotations", image_id + ".xml")
+    image_annotations = _parse_annotation_data(annotation_file_path)
+
+    result = {
+        "image/filename": image_id + ".jpg",
+        "image/file_path": image_file_path,
+        "segmentation/class/file_path": class_segmentation_file_path,
+        "segmentation/object/file_path": object_segmentation_file_path,
+    }
+    result.update(image_annotations)
+    # Labels field should be same as the 'object.label'
+    labels = list(set([o["label"] for o in result["objects"]]))
+    result["labels"] = sorted(labels)
+    return result
+
+
+def _build_metadata(data_dir, image_ids):
+    # Parallel process all the images.
+    image_file_paths = [
+        os.path.join(data_dir, "JPEGImages", i + ".jpg") for i in image_ids
+    ]
+    p = multiprocessing.Pool(10)
```

Comment:
Is there any way to use a dynamically sized threadpool here? Using a static size seems a bit machine-specific, but I suppose it's probably not really a problem?

Reply:
Not a big problem so far, I can change it to read the local cpu count, but might be a bit overkill. 

Updated a bit to not use too many thread when the data entry is less than 10.